feat: move to Locator playwright

kobenguyent opened the pull request:
## Motivation/Description of the PR
- Resolves #3677 

Applicable helpers:
- [x] Playwright

## Type of change
- [x] :bug: Bug fix


## Checklist:
- [x] Lint checking (Run `npm run lint`)
- [x] Local tests are passed (Run `npm test`)

kobenguyent committed:
feat: move to Locator playwright

kobenguyent committed:
fix: docs

kobenguyent committed:
fix: acceptance tests

kobenguyent committed:
fix: acceptance tests

kobenguyent committed:
fix: acceptance tests

kobenguyent committed:
fix: acceptance tests

DavertMik commented on lib/helper/Playwright.js:
```diff
@@ -1556,21 +1551,16 @@ class Playwright extends Helper {
-      result = await this.fillField(locator, '');
-    }
-    return result;
+    await el.fill('');
```
why we don't use `clear` in this case?
`fill('')` could produce different result

DavertMik reviewed (changes requested):
Good to me
But can we use new `clear` method?

kobenguyent committed:
Update Playwright.js

hatufacci commented on lib/helper/Playwright.js:
```diff
@@ -1526,13 +1526,8 @@ class Playwright extends Helper {
-      await this._evaluateHandeInContext(el => el.innerHTML = '', el);
-    }
+
+    await el.clear();
```
@kobenguyent 
why we need clear() ? in fillField() method ? we have separated method for this.
Do we need clear field before fill something in all cases?

kobenguyent replied:
I don't know why, it was like that to make sure the field is all clear before we fill something out.

hatufacci replied:
I'm ok with that but i prefer leave such things at the user's discretion :)

kobenguyent replied:
Understood but that behavior was like that before so just don't wanna break existing tests of users. 😊

kobenguyent committed:
merged with 3.x

kobenguyent committed:
fix: remove some leftover

kobenguyent committed:
fix: some failed tests

kobenguyent committed:
fix: some failed tests

kobenguyent committed:
fix: some failed tests

EgorBodnar commented on test/helper/webapi.js:
```diff
@@ -1121,10 +1121,10 @@ module.exports.tests = function () {
+      I._withinBegin('#register')
+        .then(() => I.see('E-Mail'))
+        .then(() => I.dontSee('Toggle navigation'))
+        .then(() => I.dontSeeElement('#navbar-collapse-menu'));
```
Could we switch here to async/await?

kobenguyent replied:
I'm not sure why, but async/await there failed the tests. That's why I updated to .then()

kobenguyent committed:
Merge branch '3.x' into feat/move-to-locator-playwright

EgorBodnar commented on lib/helper/Playwright.js:
```diff
@@ -2699,7 +2679,8 @@ async function getXPathForElement(elementHandle) {
   locator = new Locator(locator, 'css');
-  return matcher.$$(buildLocatorString(locator));
+
+  return matcher.locator(buildLocatorString(locator)).all();
```
It could be a huge issue for projects with dynamic rendering. 
[$$](https://playwright.dev/docs/api/class-elementhandle#element-handle-query-selector-all) from ElementHandle  awaits for all matching elements, but [.all()](https://playwright.dev/docs/api/class-locator#locator-all) from Locator not, it immediately returns whatever is present in the page. 

So as the function `findElements()` use as entry point for locating, it's dangerous to stay as is

DavertMik replied:
@EgorBodnar I think it is dangerous only if we would not call this methods in sequence

If we would have it like this:

```js
const els = await findElements()
await I.doSomeOtherStuff()
els[0].click()
```
the behavior of `$$` and `locator().all()` would be different
but if we found elements, picked the first one, clicked it, it should be the same

maybe there is slight difference for internal Playwright retries but I think they should stabilize execution, not cause failures

EgorBodnar replied:
I think we really have no way out. Users will have to pay more attention to waiting for the elements to load.

I suggest leaving the implementation of the `findElements()` function as @kobenguyent  proposed, but also adding a `findElement()` function that will not use the `.all()` method. And then in the Playwright helper, we should use the new `findElement()` function as much as possible, where work with a single element is required. This way we can increase stability.

@DavertMik @kobenguyent what do you think?

kobenguyent replied:
@EgorBodnar correct me if I'm wrong, when using `findElements()`, we just grab the first element from the list to interact with, so may you share what the differences when we use the approach to use `findElement()` where it is possible?

EgorBodnar replied:
[locator.all()](https://playwright.dev/docs/api/class-locator#locator-all) does not wait for elements to match the locator, but without `.all()` waits

So to make selectors more stable better to use it as often as we can without  the`.all()` method

DavertMik replied:
I'm not sure I understand all details here but here is the plan:

@kobenguyent try to follow suggestions from @EgorBodnar 

when you think the fix is ready I will **publish a beta release from this branch** and we will ask community to test it. If everything is ok, I will merge it and make a proper release

kobenguyent replied:
hey @DavertMik, updated as suggestions from @EgorBodnar

kobenguyent committed:
Merge branch '3.x' into feat/move-to-locator-playwright

kobenguyent committed:
Merge branch '3.x' into feat/move-to-locator-playwright

kobenguyent committed:
fix: use _react locator of Playwright

kobenguyent committed:
fix: try to use findElement() where possible

kobenguyent committed:
fix: failed tests

DavertMik commented:
@kobenguyent @EgorBodnar 

Published this PR as `codeceptjs@3.5.4-beta.1`
Please check this new version

kobenguyent commented:
> @kobenguyent @EgorBodnar 
> 
> 
> 
> Published this PR as `codeceptjs@3.5.4-beta.1`
> 
> Please check this new version
> 
> 

Thank you @DavertMik 😊

ngraf commented:
I ran my biggest test suite with a lot of different locators and techniques against 3.5.4-beta.1 with Playwright.
I discovered one issue:

🐞 **Issue**: `I.switchTo(locator)` step does not fail when locator was not found.
*Expected*
`switchTo` does fail when locator does not match any element.
*Actual*
`switchTo` does not fail when locator does not match any element.

Proof: 
This scenario fails as expected with latest stable version, but is unexpectedly green in tested beta version:
```
Scenario('Playwright locator Bug with iframe not found', async ({I}) => {
    I.amOnPage('https://www.w3schools.com/html/html_iframe.asp')
    // I.switchTo('iframe[ title = "W3Schools HTML Tutorial']) // This would be the correct locator
    I.switchTo('iframe[ title = "W3Schools HTML Tutorial    -> I ADDED THIS TEXT WITH THE GOAL TO NOT FIND IFRAME. TEST SHOULD FAIL BUT IT DOES NOT WITH CodeceptJS version 3.5.4-beta.1 <-"]')

})
```

kobenguyent commented:
> I.amOnPage('https://www.w3schools.com/html/html_iframe.asp')
>     // I.switchTo('iframe[ title = "W3Schools HTML Tutorial']) // This would be the correct locator
>     I.switchTo('iframe[ title = "W3Schools HTML Tutorial    -> I ADDED THIS TEXT WITH THE GOAL TO NOT FIND IFRAME. TEST SHOULD FAIL BUT IT DOES NOT WITH CodeceptJS version 3.5.4-beta.1 <-"]')

Hi @ngraf I tried your scenario against 3.x and nothing fails, so I think it's not the issue from this Locator changes.

```
    it.only('should switch reference to iframe content', () => {
      I.amOnPage('https://www.w3schools.com/html/html_iframe.asp');
      // I.switchTo('iframe[ title = "W3Schools HTML Tutorial']) // This would be the correct locator
      I.switchTo('iframe[ title = "W3Schools HTML Tutorial    -> I ADDED THIS TEXT WITH THE GOAL TO NOT FIND IFRAME. TEST SHOULD FAIL BUT IT DOES NOT WITH CodeceptJS version 3.5.4-beta.1 <-"]')

    });
```

```
➜  CodeceptJS git:(3.x) ✗ npm run test:unit:webbapi:playwright

> codeceptjs@3.5.3 test:unit:webbapi:playwright
> mocha test/helper/Playwright_test.js



  Playwright
    #switchTo
      ✔ should switch reference to iframe content


  1 passing (825ms)

➜  CodeceptJS git:(3.x) ✗ 

```

kobenguyent committed:
fix: handle the case Node is not an <input>, <textarea> or <select> element

ngraf commented:
Ok, I guess you are right and the issue is not caused by this feature branch then.

kobenguyent committed:
merged with 3.x
